### PR TITLE
Port config fix to 2.2

### PR DESCRIPTION
**Pull request by HaoK**
@ajcvickers

### CODE REVIEW

File: src/Configuration/test/Config.FunctionalTests/ConfigurationTests.cs [hunk 1]
```diff
@@ -866,6 +869,79 @@ public void CanEnumerateProviders()
             Assert.NotNull(providers.Single(p => p is IniConfigurationProvider));
         }
 
+        [ConditionalFact]
+        [OSSkipCondition(OperatingSystems.Linux, SkipReason = "File watching is flaky on non windows.")]
+        [OSSkipCondition(OperatingSystems.MacOSX, SkipReason = "File watching is flaky on non windows.")]
```
Comment (Tratcher) [inline src/Configuration/test/Config.FunctionalTests/ConfigurationTests.cs]: Remove these skips, they were also removed from the original PR.

**Comment by HaoK** [comment]
There's no reload for user secrets in 2.2 so those tests had to be removed

**Comment by ajcvickers** [comment]
Original PR: https://github.com/aspnet/Extensions/pull/1202
Issue: https://github.com/aspnet/Extensions/issues/1189

#### Description

There is a race condition that can be hit if a file is being updated while the configuration system is refreshing configuration from that file.

#### Customer Impact

Customer reported. Application crashes with exception.

#### Regression?

No; same behavior in 2.1. At this time, we're bring for 2.2 only.

#### Risk

Low. Configuration system is relatively small and code change is simple.

**Comment by martincostello** [comment]
For my own curiosity, how often does _"Shiproom"_ meet to discuss such proposed changes?

**Comment by ajcvickers** [comment]
@martincostello Usually twice per week.

**Comment by martincostello** [comment]
Thanks - I was just wondering as I was reviewing our production logs from over the weekend and we had a handful of the exception occur, so I just wanted to what the status was.

**Comment by dougbu** [comment]
This fix is incomplete without an update to [eng/PatchConfig.props](https://github.com/aspnet/AspNetCore/blob/release/2.2/eng/PatchConfig.props#L46-L49).

**Comment by vivmishra** [comment]
Approved for 2.2.4
Checkin after branding update @dougbu

**Comment by HaoK** [comment]
@dougbu what am I supposed to do to patchconfig.props?

**Comment by martincostello** [comment]
I think `Microsoft.Extensions.Configuration.KeyPerFile` and `Microsoft.Extensions.Configuration.EnvironmentVariables` need adding to the 2.2.4 node.

**Comment by HaoK** [comment]
Alright, @dougbu updated patchconfig

File: src/Configuration/test/Config.FunctionalTests/Microsoft.Extensions.Configuration.FunctionalTests.csproj [hunk 2]
```diff
@@ -5,6 +5,13 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\..\testassets\Test.Common\Microsoft.Extensions.Configuration.Test.Common.csproj" />
```
Comment (dougbu) [inline src/Configuration/test/Config.FunctionalTests/Microsoft.Extensions.Configuration.FunctionalTests.csproj]: Full path of this project is `src/Configuration/test/Config.Test.Common/Microsoft.Extensions.Configuration.Test.Common.csproj` So, this should be `..\Config.Test.Common/Microsoft.Extensions.Configuration.Test.Common.csproj`

**Comment by HaoK** [comment]
Hrm build checks are red after updating the patch config...  any ideas @dougbu ?

```
WARNING: 0 error(s) and 2 warning(s) found with package Microsoft.Extensions.Configuration.EnvironmentVariables (2.2.4-servicing-19164-12)
WARNING: ASSEMBLY_FILE_VERSION_MISMATCH @ lib/netstandard2.0/Microsoft.Extensions.Configuration.EnvironmentVariables.dll: The managed assembly file version '2.2.0.19073' does not match the package version '2.2.4-servicing-19164-12' for package Microsoft.Extensions.Configuration.EnvironmentVariables.
WARNING: ASSEMBLY_VERSION_MISMATCH @ lib/netstandard2.0/Microsoft.Extensions.Configuration.EnvironmentVariables.dll: The managed assembly version '2.2.0.0' does not match the package version '2.2.4.0' for package Microsoft.Extensions.Configuration.EnvironmentVariables.
INFO: Analyzing Microsoft.Extensions.Configuration.KeyPerFile (2.2.4-servicing-19164-12)
WARNING: 0 error(s) and 2 warning(s) found with package Microsoft.Extensions.Configuration.KeyPerFile (2.2.4-servicing-19164-12)
WARNING: ASSEMBLY_FILE_VERSION_MISMATCH @ lib/netstandard2.0/Microsoft.Extensions.Configuration.KeyPerFile.dll: The managed assembly file version '2.2.0.19073' does not match the package version '2.2.4-servicing-19164-12' for package Microsoft.Extensions.Configuration.KeyPerFile.
WARNING: ASSEMBLY_VERSION_MISMATCH @ lib/netstandard2.0/Microsoft.Extensions.Configuration.KeyPerFile.dll: The managed assembly version '2.2.0.0' does not match the package version '2.2.4.0' for package Microsoft.Extensions.Configuration.KeyPerFile.
WARNING: Unnecessary exclusion in Microsoft.Extensions.Logging.Testing
Issue: BUILD_ITEMS_FRAMEWORK
Instance: *

Showing JSON for ignore content:
{
dotnet.exe failed with exit code: 1
  "Microsoft.Extensions.Configuration.EnvironmentVariables": {
At D:\a\.dotnet\buildtools\korebuild\2.2.1-build-20190104.5\scripts\common.psm1:11 char:9
+         throw "$cmdName failed with exit code: $exitCode"
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (dotnet.exe failed with exit code: 1:String) [], RuntimeException
    + FullyQualifiedErrorId : dotnet.exe failed with exit code: 1
 
    "Exclusions": {
      "ASSEMBLY_FILE_VERSION_MISMATCH": {
        "lib/netstandard2.0/Microsoft.Extensions.Configuration.EnvironmentVariables.dll": "Enter justification"
      },
      "ASSEMBLY_VERSION_MISMATCH": {
        "lib/netstandard2.0/Microsoft.Extensions.Configuration.EnvironmentVariables.dll": "Enter justification"
      }
    }
  },
  "Microsoft.Extensions.Configuration.KeyPerFile": {
    "Exclusions": {
      "ASSEMBLY_FILE_VERSION_MISMATCH": {
        "lib/netstandard2.0/Microsoft.Extensions.Configuration.KeyPerFile.dll": "Enter justification"
      },
      "ASSEMBLY_VERSION_MISMATCH": {
        "lib/netstandard2.0/Microsoft.Extensions.Configuration.KeyPerFile.dll": "Enter justification"
      }
    }
  }
}
```

**Comment by dougbu** [comment]
Oh well, this isn't _quite_ ready to go

File: src/Configuration/test/Config.FunctionalTests/Microsoft.Extensions.Configuration.FunctionalTests.csproj (hunk 2 above)
Comment (HaoK) [reply src/Configuration/test/Config.FunctionalTests/Microsoft.Extensions.Configuration.FunctionalTests.csproj]: Ok i'll give that a shot

**Comment by dougbu** [comment]
@HaoK will get back to you on the version mismatches once I get home. Things aren't working as they should&hellip;

**Comment by dougbu** [comment]
@HaoK didn't find similar changes elsewhere,. But, I'm guessing the package verifier is correct.

Add the following to the `Default/packages` array in eng/NuGetPackageVerifier.json:
``` json
  "Microsoft.Extensions.Configuration.EnvironmentVariables": {
    "Exclusions": {
      "ASSEMBLY_FILE_VERSION_MISMATCH": {
        "lib/netstandard2.0/Microsoft.Extensions.Configuration.EnvironmentVariables.dll": "Enter justification"
      },
      "ASSEMBLY_VERSION_MISMATCH": {
        "lib/netstandard2.0/Microsoft.Extensions.Configuration.EnvironmentVariables.dll": "Enter justification"
      }
    }
  },
  "Microsoft.Extensions.Configuration.KeyPerFile": {
    "Exclusions": {
      "ASSEMBLY_FILE_VERSION_MISMATCH": {
        "lib/netstandard2.0/Microsoft.Extensions.Configuration.KeyPerFile.dll": "Enter justification"
      },
      "ASSEMBLY_VERSION_MISMATCH": {
        "lib/netstandard2.0/Microsoft.Extensions.Configuration.KeyPerFile.dll": "Enter justification"
      }
    }
  }
```

**Comment by HaoK** [comment]
Cool that was it, green now @dougbu

**Pull request merged**